feg/gateway/sbi/sbi_log_middleware.go

```diff
@@ -0,0 +1,164 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sbi
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/labstack/echo/v4"
+)
+
+type Logger interface {
+	// Skipper returns true to skip the logging
+	Skipper() bool
+	// LogRequest logs http request relateed info before making the client request
+	LogRequest(method string, url *url.URL, reqBody []byte, header http.Header)
+	// LogResponse logs http response related info after receiving the response from the server
+	LogResponse(url *url.URL, status string, resBody []byte, header http.Header, latency time.Duration)
+}
+
+// LoggingTransport implements the http.RoundTripper interface and logs the
+// HTTP request and response information
+type LoggingTransport struct {
+	Transport http.RoundTripper
+	Logger    Logger
+}
+
+type responseGrabber struct {
+	io.Writer
+	http.ResponseWriter
+}
+
+func NewLoggingTransport(logger Logger) LoggingTransport {
+	if logger == nil {
+		logger = &DefaultGlogLogger{}
+	}
+	return LoggingTransport{
+		Transport: http.DefaultTransport,
+		Logger:    logger,
+	}
+}
+
+// RoundTrip overrides the http.DefaultTransport RoundTrip behavior and logs the HTTP request and response.
+func (lt LoggingTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
+	if lt.Logger.Skipper() {
+		// No logging required
+		return lt.Transport.RoundTrip(req)
+	}
+
+	// Before sending
+	var reqBody []byte
+	if req.Body != nil {
+		reqBody, err = ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		// reset the req.Body back to original
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
+	}
+	lt.Logger.LogRequest(req.Method, req.URL, reqBody, req.Header)
+
+	// Make the actual request
+	start := time.Now()
+	res, err = lt.Transport.RoundTrip(req)
+	took := time.Since(start)
+	if err != nil {
+		// error responses shall be logged by caller
+		return
+	}
+
+	// After sending
+	var resBody []byte
+	if res.Body != nil {
+		resBody, err = ioutil.ReadAll(res.Body)
+		if err != nil {
+			return
+		}
+		// reset the res.Body back to original
+		res.Body = ioutil.NopCloser(bytes.NewBuffer(resBody))
+	}
+	lt.Logger.LogResponse(req.URL, res.Status, resBody, res.Header, took)
+	return
+}
+
+func (w *responseGrabber) WriteHeader(code int) {
+	w.ResponseWriter.WriteHeader(code)
+}
+
+func (w *responseGrabber) Write(b []byte) (int, error) {
+	return w.Writer.Write(b)
+}
+
+func (w *responseGrabber) Flush() {
+	w.ResponseWriter.(http.Flusher).Flush()
+}
+
+func (w *responseGrabber) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	return w.ResponseWriter.(http.Hijacker).Hijack()
+}
```

Comment:
Not sure if all those functions are just wrappers or implementation of ResponseWriter methods, but if they do not add any extra logic, you don't need them.  

Since `responseGrabber` embeds those two structs, `responseGrabber` already has those methods for free. You can try to access them if you remove those functions.

```
type responseGrabber struct {
	io.Writer
	http.ResponseWriter
}
```

https://yourbasic.org/golang/inheritance-object-oriented/

Reply:
This code is based on [BodyDumpMiddleware](https://github.com/labstack/echo/blob/4fffee2ec8a4efe5cca66cdfe17e6eeec59df60a/middleware/body_dump.go#L93).

Removing those code results in a compilation error, since bothe io.Writer and http.ResponseWriter are interfaces and not structs.